sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataCoder.java

```diff
@@ -46,14 +54,18 @@ public void encode(Metadata value, OutputStream os) throws IOException {
 
   @Override
   public Metadata decode(InputStream is) throws IOException {
+    return decodeBuilder(is).build();
+  }
+
+  Metadata.Builder decodeBuilder(InputStream is) throws IOException {
     ResourceId resourceId = RESOURCE_ID_CODER.decode(is);
     boolean isReadSeekEfficient = INT_CODER.decode(is) == 1;
     long sizeBytes = LONG_CODER.decode(is);
     return Metadata.builder()
         .setResourceId(resourceId)
         .setIsReadSeekEfficient(isReadSeekEfficient)
         .setSizeBytes(sizeBytes)
-        .build();
+        .setLastModifiedMillis(UNKNOWN_LAST_MODIFIED_MILLIS);
```

Comment:
The idea here is that lastModifiedMillis must have some default value. I've chosen here to use `-1` so that `lastModifiedMillis` doesn't have to be nullable, though we can discuss if null makes more sense here.  

AutoValue requires that values be set for all members. To define a default value, the AutoValue docs suggest a pattern of calling setters before returning the builder. That's what's going on here.

I don't see that there's a backwards incompatibility here. MetadataCoder still only writes and reads the three existing values (resource id, int, and long). When encoding, it throws away lastModifiedMills and when decoding, it provides the default -1 value.

Do you have thoughts on whether `null` would be preferred as the default vs. `-1`?

Reply:
Of note, Java's `File.lastModified` documents that `0` is used if the file does not exist or an I/O error occurs. We could follow that and use `0`.

Reply:
What if we let zero as a default. In that case we won't need that and we will still be ok with AutoValue. If we setup any default I think it is better to do it at the core object level `Metadata.create()`  better than at other places. Also in the current case if you put this only in `decode` then encode won't put the right value. Better to left this class as it was to avoid the risk of breaking stuff..

Reply:
> What if we let zero as a default. In that case we won't need that and we will still be ok with AutoValue

To be clear, `0` is only used as a default for `File.lastModified` as used in `LocalFileSystem`. Other filesystems do not have this same concept of 0 as default. So any way we go, we will have to explicitly set some default value for the field.

I agree with you that it's better to set the default within `Metadata` rather than only in the coder, so I will make that change. And will also change the default to be 0 to be consistent with `File.lastModified` behavior.

> Also in the current case if you put this only in decode then encode won't put the right value

Exactly. `encode` _does not_ encode the lastModifiedMillis value; it's thrown away. And decode provides a default. So encode and decode remain matched in the number and types of encoded values, and these are compatible with previous beam versions.